### PR TITLE
8298385: Some font classes rely on blind casting to implement Object.equals()

### PR DESCRIPTION
**Pull request by SWinxy**
Some also had two separate methods to check equality, one calling the other. Objects.equals() is used in some places as a replacement for a local copy of the function. In the end, the equals methods become quicker to understand their checks, more condensed, and becomes more in line with other equality methods in the JDK.

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8298385](https://bugs.openjdk.org/browse/JDK-8298385): Some font classes rely on blind casting to implement Object.equals() ⚠️ Issue is not open.

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10416/head:pull/10416` \
`$ git checkout pull/10416`

Update a local copy of the PR: \
`$ git checkout pull/10416` \
`$ git pull https://git.openjdk.org/jdk pull/10416/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10416`

View PR using the GUI difftool: \
`$ git pr show -t 10416`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10416.diff">https://git.openjdk.org/jdk/pull/10416.diff</a>

</details>

### CODE REVIEW

File: src/java.desktop/share/classes/java/awt/font/FontRenderContext.java [hunk 1]
```diff
@@ -290,51 +291,29 @@ public Object getFractionalMetricsHint() {
...
+            // if neither instance is a subclass, reference values directly.
+            if (!o1.defaulting && !defaulting) {
+                return o1.aaHintValue == aaHintValue
+                        && o1.fmHintValue == fmHintValue
+                        && Objects.equals(tx, o1.tx);
+            } else {
+                return o1.getAntiAliasingHint() == getAntiAliasingHint()
+                        && o1.getFractionalMetricsHint() == getFractionalMetricsHint()
+                        && o1.getTransform().equals(getTransform());
             }
```
Comment (ExE-Boss) [inline src/java.desktop/share/classes/java/awt/font/FontRenderContext.java:314]:
```suggestion
        if (o instanceof FontRenderContext rhs) {
            // if neither instance is a subclass, reference values directly.
            if (!rhs.defaulting && !defaulting) {
                return rhs.aaHintValue == aaHintValue
                        && rhs.fmHintValue == fmHintValue
                        && Objects.equals(tx, rhs.tx);
            } else {
                return rhs.getAntiAliasingHint() == getAntiAliasingHint()
                        && rhs.getFractionalMetricsHint() == getFractionalMetricsHint()
                        && rhs.getTransform().equals(getTransform());
            }
```

File: src/java.desktop/share/classes/java/awt/font/ImageGraphicAttribute.java [hunk 2]
```diff
@@ -176,52 +176,17 @@ public int hashCode() {
...
-
-        if (!fImage.equals(rhs.fImage)) {
-            return false;
-        }
+    public boolean equals(Object o) {
+        return this == o || o instanceof ImageGraphicAttribute o1
+                && fOriginX == o1.fOriginX
+                && fOriginY == o1.fOriginY
+                && getAlignment() == o1.getAlignment()
+                && fImage.equals(o1.fImage);
```
Comment (ExE-Boss) [inline src/java.desktop/share/classes/java/awt/font/ImageGraphicAttribute.java:190]:
```suggestion
        return this == o || o instanceof ImageGraphicAttribute rhs
                && fOriginX == rhs.fOriginX
                && fOriginY == rhs.fOriginY
                && getAlignment() == rhs.getAlignment()
                && fImage.equals(rhs.fImage);
```

File: src/java.desktop/share/classes/java/awt/font/ShapeGraphicAttribute.java [hunk 3]
```diff
@@ -205,52 +205,15 @@ public int hashCode() {
...
-        if (!fShape.equals(rhs.fShape)) {
-            return false;
-        }
-
-        return true;
+    public boolean equals(Object o) {
+        return this == o || o instanceof ShapeGraphicAttribute o1
+                && fStroke == o1.fStroke
+                && getAlignment() == o1.getAlignment()
+                && fShape.equals(o1.fShape);
```
Comment (ExE-Boss) [inline src/java.desktop/share/classes/java/awt/font/ShapeGraphicAttribute.java:217]:
```suggestion
        return this == o || o instanceof ShapeGraphicAttribute rhs
                && fStroke == rhs.fStroke
                && getAlignment() == rhs.getAlignment()
                && fShape.equals(rhs.fShape);
```

File: src/java.desktop/share/classes/java/awt/font/TransformAttribute.java [hunk 4]
```diff
@@ -151,24 +151,14 @@ public int hashCode() {
...
-                }
-                return transform.equals(that.transform);
-            }
-            catch (ClassCastException e) {
-            }
-        }
-        return false;
+    public boolean equals(Object o) {
+        return this == o || o instanceof TransformAttribute o1
+                && transform.equals(o1.transform);
```
Comment (ExE-Boss) [inline src/java.desktop/share/classes/java/awt/font/TransformAttribute.java:162]:
```suggestion
        return this == o || o instanceof TransformAttribute rhs
                && transform.equals(rhs.transform);
```

File: src/java.desktop/share/classes/sun/font/AttributeValues.java [hunk 5]
```diff
@@ -436,48 +437,38 @@ public int hashCode() {
...
+                && Objects.equals(numericShaping, o1.numericShaping)
+                && Objects.equals(justification, o1.justification)
+                && Objects.equals(charReplacement, o1.charReplacement)
+                && size == o1.size
+                && weight == o1.weight
+                && posture == o1.posture
+                && Objects.equals(family, o1.family)
+                && Objects.equals(font, o1.font)
+                && imUnderline == o1.imUnderline
+                && Objects.equals(imHighlight, o1.imHighlight);
```
Comment (ExE-Boss) [inline src/java.desktop/share/classes/sun/font/AttributeValues.java:471]:
```suggestion
        return this == o || o instanceof AttributeValues rhs
                && defined == rhs.defined
                && nondefault == rhs.nondefault
                && underline == rhs.underline
                && strikethrough == rhs.strikethrough
                && superscript == rhs.superscript
                && width == rhs.width
                && kerning == rhs.kerning
                && tracking == rhs.tracking
                && ligatures == rhs.ligatures
                && runDirection == rhs.runDirection
                && bidiEmbedding == rhs.bidiEmbedding
                && swapColors == rhs.swapColors
                && Objects.equals(transform, rhs.transform)
                && Objects.equals(foreground, rhs.foreground)
                && Objects.equals(background, rhs.background)
                && Objects.equals(numericShaping, rhs.numericShaping)
                && Objects.equals(justification, rhs.justification)
                && Objects.equals(charReplacement, rhs.charReplacement)
                && size == rhs.size
                && weight == rhs.weight
                && posture == rhs.posture
                && Objects.equals(family, rhs.family)
                && Objects.equals(font, rhs.font)
                && imUnderline == rhs.imUnderline
                && Objects.equals(imHighlight, rhs.imHighlight);
```

File: src/java.desktop/share/classes/sun/font/CoreMetrics.java [hunk 6]
```diff
@@ -70,36 +70,21 @@ public int hashCode() {
...
+                && baselineIndex == o1.baselineIndex
+                && baselineOffsets[0] == o1.baselineOffsets[0]
+                && baselineOffsets[1] == o1.baselineOffsets[1]
+                && baselineOffsets[2] == o1.baselineOffsets[2]
+                && strikethroughOffset == o1.strikethroughOffset
+                && strikethroughThickness == o1.strikethroughThickness
+                && underlineOffset == o1.underlineOffset
+                && underlineThickness == o1.underlineThickness
+                && ssOffset == o1.ssOffset
+                && italicAngle == o1.italicAngle;
```
Comment (ExE-Boss) [inline src/java.desktop/share/classes/sun/font/CoreMetrics.java:87]:
```suggestion
        return this == o || o instanceof CoreMetrics rhs
                && ascent == rhs.ascent
                && descent == rhs.descent
                && leading == rhs.leading
                && baselineIndex == rhs.baselineIndex
                && baselineOffsets[0] == rhs.baselineOffsets[0]
                && baselineOffsets[1] == rhs.baselineOffsets[1]
                && baselineOffsets[2] == rhs.baselineOffsets[2]
                && strikethroughOffset == rhs.strikethroughOffset
                && strikethroughThickness == rhs.strikethroughThickness
                && underlineOffset == rhs.underlineOffset
                && underlineThickness == rhs.underlineThickness
                && ssOffset == rhs.ssOffset
                && italicAngle == rhs.italicAngle;
```

File: src/java.desktop/share/classes/sun/font/Decoration.java [hunk 7]
```diff
@@ -168,48 +165,14 @@ private static final class DecorationImpl extends Decoration {
...
-            }
-            return areEqual(imUnderline, other.imUnderline);
+        public boolean equals(Object o) {
+            return this == o || o instanceof DecorationImpl o1
+                    && swapColors == o1.swapColors
+                    && strikethrough == o1.strikethrough
+                    && Objects.equals(stdUnderline, o1.stdUnderline)
+                    && Objects.equals(fgPaint, o1.fgPaint)
+                    && Objects.equals(bgPaint, o1.bgPaint)
+                    && Objects.equals(imUnderline, o1.imUnderline);
```
Comment (ExE-Boss) [inline src/java.desktop/share/classes/sun/font/Decoration.java:175]:
```suggestion
            return this == o || o instanceof DecorationImpl other
                    && swapColors == other.swapColors
                    && strikethrough == other.strikethrough
                    && Objects.equals(stdUnderline, other.stdUnderline)
                    && Objects.equals(fgPaint, other.fgPaint)
                    && Objects.equals(bgPaint, other.bgPaint)
                    && Objects.equals(imUnderline, other.imUnderline);
```

File: src/java.desktop/share/classes/sun/font/FontLineMetrics.java [hunk 8]
```diff
@@ -98,13 +99,8 @@ public int hashCode() {
...
 
-    public boolean equals(Object rhs) {
-        try {
-            return cm.equals(((FontLineMetrics)rhs).cm);
-        }
-        catch (ClassCastException e) {
-            return false;
-        }
+    public boolean equals(Object o) {
+        return this == o || o instanceof FontLineMetrics o1 && Objects.equals(cm, o1.cm);
```
Comment (ExE-Boss) [inline src/java.desktop/share/classes/sun/font/FontLineMetrics.java:103]:
```suggestion
        return this == o || o instanceof FontLineMetrics rhs
            && Objects.equals(cm, rhs.cm);
```

File: src/java.desktop/share/classes/sun/font/GlyphLayout.java [hunk 9]
```diff
@@ -134,18 +134,11 @@ int lang() {
...
-                       this.font.equals(that.font);
-            }
-            catch (ClassCastException e) {
-                return false;
-            }
+        public boolean equals(Object o) {
+            return this == o || o instanceof LayoutEngineKey o1
+                    && script == o1.script
+                    && lang == o1.lang
+                    && font.equals(o1.font);
```
Comment (ExE-Boss) [inline src/java.desktop/share/classes/sun/font/GlyphLayout.java:141]:
```suggestion
            return this == o || o instanceof LayoutEngineKey rhs
                    && script == rhs.script
                    && lang == rhs.lang
                    && font.equals(rhs.font);
```

File: src/java.desktop/share/classes/sun/font/GlyphLayout.java [hunk 10]
```diff
@@ -279,16 +272,10 @@ public int hashCode() {
...
-                        font.equals(rhs.font) &&
-                        frc.equals(rhs.frc);
-                }
-                catch (ClassCastException e) {
-                }
-                return false;
+                return this == o || o instanceof SDKey o1
+                        && hash == o1.hash
+                        && font.equals(o1.font)
+                        && frc.equals(o1.frc);
```
Comment (ExE-Boss) [inline src/java.desktop/share/classes/sun/font/GlyphLayout.java:278]:
```suggestion
                return this == o || o instanceof SDKey rhs
                        && hash == rhs.hash
                        && font.equals(rhs.font)
                        && frc.equals(rhs.frc);
```

File: src/java.desktop/share/classes/sun/font/StandardGlyphVector.java [hunk 11]
```diff
@@ -634,83 +635,65 @@ public GlyphJustificationInfo getGlyphJustificationInfo(int ix) {
...
-     * the inherited Object.equals(Object) as well.  GlyphVector should do
-     * this, and define two glyphvectors as not equal if the classes differ.
-     */
-    public boolean equals(Object rhs) {
-        try {
-            return equals((GlyphVector)rhs);
-        }
-        catch (ClassCastException e) {
-            return false;
+            return Objects.equals(gti, o1.gti);
```
Comment (ExE-Boss) [inline src/java.desktop/share/classes/sun/font/StandardGlyphVector.java:694]:
```suggestion
        if (o instanceof StandardGlyphVector other) {
            if (glyphs.length != other.glyphs.length) {
                return false;
            }

            for (int i = 0; i < glyphs.length; ++i) {
                if (glyphs[i] != other.glyphs[i]) {
                    return false;
                }
            }

            if (!font.equals(other.font)) {
                return false;
            }

            if (!frc.equals(other.frc)) {
                return false;
            }

            if ((other.positions == null) != (positions == null)) {
                if (positions == null) {
                    initPositions();
                } else {
                    other.initPositions();
                }
            }

            if (positions != null) {
                for (int i = 0; i < positions.length; ++i) {
                    if (positions[i] != other.positions[i]) {
                        return false;
                    }
                }
            }

            return Objects.equals(gti, other.gti);
```

**Review by ExE-Boss — COMMENTED** [review]
Maybe keep the old variable names:

**Comment by SWinxy** [comment]
@ExE-Boss are you in the JBS? If you are, would you please add a bug report for this PR?

**Comment by ExE-Boss** [comment]
@SWinxy
> @ExE-Boss are you in the JBS?

No, I’m not.

> If you are, would you please add a bug report for this PR?

You can use https://bugreport.java.com/bugreport/ to add it yourself.

File: src/java.desktop/share/classes/sun/font/PhysicalFont.java [hunk 12]
```diff
@@ -41,14 +42,9 @@ public abstract class PhysicalFont extends Font2D {
...
     public boolean equals(Object o) {
-        if (o == null || o.getClass() != this.getClass()) {
-            return false;
-        }
-        PhysicalFont other = (PhysicalFont)o;
-        return
-           (this.fullName.equals(other.fullName)) &&
-            ((this.platName == null && other.platName == null) ||
-             (this.platName != null && this.platName.equals(other.platName)));
+        return o instanceof PhysicalFont other
```
Comment (ExE-Boss) [inline src/java.desktop/share/classes/sun/font/PhysicalFont.java:45]: The old implementation also checked that the runtime class of `other` is the same as that of `this`:
```suggestion
        return o instanceof PhysicalFont other
                && other.getClass() == this.getClass()
```

Comment (SWinxy) [reply src/java.desktop/share/classes/sun/font/PhysicalFont.java:45]: Intentional change. There is only one class that extends PhysicalFont, which also happens to overrides PhysicalFont's equals. So checking via instanceof shouldn't mean any change. Also I think I forgot to add a `o == this ||` to some of them... Tomorrow's problem.

**Comment by TheShermanTanker** [comment]
Sorry for being so late, here's the corresponding entry: https://bugs.openjdk.org/browse/JDK-8298385

**Comment by SWinxy** [comment]
Looks like there are issues with my PR that I haven't noticed. Shoot.

File: src/java.desktop/share/classes/java/awt/font/FontRenderContext.java [hunk 13]
```diff
@@ -290,51 +291,29 @@ public Object getFractionalMetricsHint() {
...
-    /**
-     * Return true if rhs has the same transform, antialiasing,
-     * and fractional metrics values as this.
-     * @param rhs the {@code FontRenderContext} to test for equality
-     * @return {@code true} if {@code rhs} is equal to
-     *         this {@code FontRenderContext}; {@code false}
-     *         otherwise.
-     * @since 1.4
-     */
-    public boolean equals(FontRenderContext rhs) {
```
Comment (mrserb) [inline src/java.desktop/share/classes/java/awt/font/FontRenderContext.java:316]: The public methods in the public classes in the `java/awt/*` package are part of the public API, it is not possible to delete such methods.

**Comment by mrserb** [comment]
> Looks like there are issues with my PR that I haven't noticed. Shoot.

While you are working on it you can move the PR to the draft state.

File: src/java.desktop/share/classes/java/awt/font/FontRenderContext.java (hunk 13 above)
Comment (SWinxy) [reply src/java.desktop/share/classes/java/awt/font/FontRenderContext.java:316]: Whoops. Thanks for catching that.

**Comment by SWinxy** [comment]
I forgot to re-apply the bug ID when I fixed the issues. Checks fail because of build chain, and GitHub isn't letting me re-run the jobs. Grr.

**Comment by SWinxy** [comment]
@mrserb would you kindly give feedback? (Linux builds failing because they feel like it.)

**Comment by mrserb** [comment]
You can restart the presubmit tests by merging from master to make it green.

**Comment by prrace** [comment]
Whoa ! This is a lot of changes for what benefit ? 
To introduce changes in behaviour in code that has been stable for 20 years ?
Every single one of these changes will need very careful analysis and testing.
Frankly, I'm not very interested in taking changes like this.

**Comment by SWinxy** [comment]
Gah. Alright.

**Pull request closed without merging**